File: Binary.java
/**
 * Binary
 */
public class Binary {

  public static int ToDecimal(String binary) {
    if (Binary.contains(".")) {
      String FractionPart = Binary.substring(Binary.indexOf('.') + 1);
      String IntegerPart = Binary.replace("." + FractionPart, "");
      int num1 = 0;
      char[] IntegeR = IntegerPart.toCharArray();
      int counter1 = 0;
      for (int i = IntegerPart.length() - 1; i >= 0; i--) {
        int a = (IntegeR[i] - '0');
        num1 += (a * Math.pow(2, counter1));
        counter1++;
      }
      String Dicemal1 = Integer.toString(num1);

      Double num2 = 0.0;
      char[] FractioN = FractionPart.toCharArray();
      int counter2 = -1;
      for (int i = 0; i < FractionPart.length(); i++) {
        int a = (FractioN[i] - '0');
        num2 += (a * Math.pow(2, counter2));
        counter2--;
      }
      String Dicemal2 = Double.toString(num2);
      String Dicemal = Dicemal1 + "." + Dicemal2.substring(2);
      return Dicemal;
    } else {
      String IntegerPart = Binary;
      int num1 = 0;
      char[] IntegeR = IntegerPart.toCharArray();
      int counter1 = 0;
      for (int i = IntegerPart.length() - 1; i >= 0; i--) {
        int a = (IntegeR[i] - '0');
        num1 += (a * Math.pow(2, counter1));
        counter1++;
      }
      String Dicemal = Integer.toString(num1);
      return Dicemal;
    }
  }

  public static String ToOctal(String binary) {
    if (Binary.contains(".")) {
      String FractionsPart = Binary.substring(Binary.indexOf('.') + 1);
      String IntegerPart = Binary.replace("." + FractionsPart, "");
      String I;
      while (IntegerPart.length() % 3 != 0) {
        I = "0" + IntegerPart;
        IntegerPart = I;
      }
      while (FractionsPart.length() % 3 != 0) {
        I = FractionsPart + "0";
        FractionsPart = I;
      }
      String Zeros = "";
      int i = 0;
      while (FractionsPart.substring(i, i + 3).equals("000")) {
        Zeros = Zeros.concat("0");
        i += 3;
        if ((i + 3) > FractionsPart.length()) {
          break;
        }
      }

      int num1 = (Integer.parseInt(IntegerPart, 2));
      String Octal1 = Integer.toOctalString(num1);
      int num2 = (Integer.parseInt(FractionsPart, 2));
      String Octal2 = Integer.toOctalString(num2);

      while (Octal2.charAt(Octal2.length() - 1) == '0') {
        Octal2 = removeLastCharacter(Octal2);
      }

      String Octal = Octal1 + "." + Zeros + Octal2;
      return Octal;
    } else {
      String I;
      while (Binary.length() % 3 != 0) {
        I = "0" + Binary;
        Binary = I;
      }
      int num = (Integer.parseInt(Binary, 2));
      String Octal = Integer.toOctalString(num);
      return Octal;
    }
  }

  public static String ToHexadecimal(String binary) {
    if (Binary.contains(".")) {
      String FractionsPart = Binary.substring(Binary.indexOf('.') + 1);
      String IntegerPart = Binary.replace("." + FractionsPart, "");
      String I;
      while (IntegerPart.length() % 4 != 0) {
        I = "0" + IntegerPart;
        IntegerPart = I;
      }
      while (FractionsPart.length() % 4 != 0) {
        I = FractionsPart + "0";
        FractionsPart = I;
      }

      String Zeros = "";
      int i = 0;
      while (FractionsPart.substring(i, i + 4).equals("0000")) {
        Zeros = Zeros.concat("0");
        i += 4;
        if ((i + 4) > FractionsPart.length()) {
          break;
        }
      }
      int num1 = (Integer.parseInt(IntegerPart, 2));
      String Hex1 = Integer.toHexString(num1);
      int num2 = (Integer.parseInt(FractionsPart, 2));
      String Hex2 = Integer.toHexString(num2);

      while (Hex2.charAt(Hex2.length() - 1) == '0') {
        Hex2 = removeLastCharacter(Hex2);
      }

      String Hex = Hex1.toUpperCase() + "." + Zeros + Hex2.toUpperCase();
      return Hex;
    } else {
      String I;
      while (Binary.length() % 4 != 0) {
        I = "0" + Binary;
        Binary = I;
      }

      int num = (Integer.parseInt(Binary, 2));
      String Hex = Integer.toHexString(num).toUpperCase();
      return Hex;
    }
  }
}

File: lib/models/binary.dart
import 'dart:math' as math;

class Binary {
  static String toDecimal(String binary) {
    if (binary.contains(".")) {
      String fractionPart = binary.substring(binary.indexOf('.') + 1);
      String integerPart = binary.replaceFirst(".$fractionPart", "");
      int num1 = 0;
      List<int> integerDigits = integerPart.split('').map(int.parse).toList();
      int counter1 = 0;
      for (int i = integerPart.length - 1; i >= 0; i--) {
        int a = integerDigits[i];
        num1 += (a * math.pow(2, counter1)).toInt();
        counter1++;
      }
      String decimal1 = num1.toString();

      double num2 = 0.0;
      List<int> fractionDigits = fractionPart.split('').map(int.parse).toList();
      int counter2 = -1;
      for (int i = 0; i < fractionPart.length; i++) {
        int a = fractionDigits[i];
        num2 += (a * math.pow(2, counter2));
        counter2--;
      }
      String decimal2 = num2.toString();
      String decimal = "$decimal1.${decimal2.substring(2)}";
      return decimal;
    } else {
      String integerPart = binary;
      int num1 = 0;
      List<int> integerDigits = integerPart.split('').map(int.parse).toList();
      int counter1 = 0;
      for (int i = integerPart.length - 1; i >= 0; i--) {
        int a = integerDigits[i];
        num1 += (a * math.pow(2, counter1)).toInt();
        counter1++;
      }
      String decimal = num1.toString();
      return decimal;
    }
  }

  static String toOctal(String binary) {
    if (binary.contains(".")) {
      String fractionsPart = binary.substring(binary.indexOf('.') + 1);
      String integerPart = binary.replaceFirst(".$fractionsPart", "");
      String i;
      while (integerPart.length % 3 != 0) {
        i = "0$integerPart";
        integerPart = i;
      }
      while (fractionsPart.length % 3 != 0) {
        i = "${fractionsPart}0";
        fractionsPart = i;
      }
      String zeros = "";
      int j = 0;
      while (fractionsPart.substring(j, j + 3) == "000") {
        zeros = "${zeros}0";
        j += 3;
        if ((j + 3) > fractionsPart.length) {
          break;
        }
      }

      int num1 = int.parse(integerPart, radix: 2);
      String octal1 = num1.toRadixString(8);
      int num2 = int.parse(fractionsPart, radix: 2);
      String octal2 = num2.toRadixString(8);

      while (octal2[octal2.length - 1] == '0') {
        octal2 = octal2.substring(0, octal2.length - 1);
      }

      String octal = "$octal1.$zeros$octal2";
      return octal;
    } else {
      String i;
      while (binary.length % 3 != 0) {
        i = "0$binary";
        binary = i;
      }
      int num = int.parse(binary, radix: 2);
      String octal = num.toRadixString(8);
      return octal;
    }
  }

  static String removeLastCharacter(String str) {
    if (str.isEmpty) {
      return str;
    }
    return str.substring(0, str.length - 1);
  }

  static String toHexadecimal(String binary) {
    bool isNegative = binary.startsWith("-");
    if (isNegative) {
      binary = binary.substring(1);
    }

    String fractionsPart = "";
    String integerPart = binary;

    if (binary.contains(".")) {
      fractionsPart = binary.substring(binary.indexOf('.') + 1);
      integerPart = binary.substring(0, binary.indexOf('.'));
    }

    // Pad IntegerPart to have a length multiple of 4
    integerPart = padToMultipleOfFour(integerPart);

    // Convert IntegerPart to hexadecimal
    String hex1 =
        int.parse(integerPart, radix: 2).toRadixString(16).toUpperCase();

    String hex2 = "";
    if (fractionsPart.isNotEmpty) {
      // Convert fractionsPart to decimal value first
      double fractionalDecimal = 0.0;
      for (int i = 0; i < fractionsPart.length; i++) {
        if (fractionsPart[i] == '1') {
          fractionalDecimal += 1 / math.pow(2, i + 1);
        }
      }

      // Convert decimal fraction to hexadecimal
      int fractionalInt;
      for (int i = 0; i < 10 && fractionalDecimal > 0; i++) {
        // Limit the conversion to 10 digits for precision
        fractionalDecimal *= 16;
        fractionalInt = fractionalDecimal.floor();
        hex2 += fractionalInt.toRadixString(16).toUpperCase();
        fractionalDecimal -= fractionalInt;
      }
    }

    // Remove trailing zeros from hex2
    hex2 = hex2.replaceAll(RegExp(r'0+$'), '');

    // Construct the final hexadecimal string
    String hex = hex2.isNotEmpty ? "$hex1.$hex2" : hex1;
    if (isNegative) {
      hex = "-$hex";
    }
    return hex;
  }

  // Helper method to pad a binary string to have a length multiple of 4
  static String padToMultipleOfFour(String binary) {
    while (binary.length % 4 != 0) {
      binary = "0$binary";
    }
    return binary;
  }
}
